src/lib.rs

```diff
@@ -333,29 +348,56 @@ impl<T: Ord> OrderedCollection<T> {
+        let prefetch_ptr = self.items.as_ptr().wrapping_add(Self::OFFSET);
 
         while i < self.items.len() {
-            let offset = (Self::MULTIPLIER * i + Self::OFFSET) & mask;
-            do_prefetch(self.items.as_ptr().wrapping_add(offset));
+            let offset = (Self::MULTIPLIER * i) & mask;
+            do_prefetch(prefetch_ptr.wrapping_add(offset));
 
-            // safe because i < self.items.len()
-            let value = unsafe { self.items.get_unchecked(i) }.borrow();
+            // safe because 1 <= i < self.items.len()
+            let value = unsafe { self.items.get_unchecked(i).assume_init_ref() }.borrow();
             // using branchless index update. At the moment compiler cannot reliably tranform
             // if expressions to branchless instructions like `cmov` and `setb`
-            i = 2 * i + 1 + usize::from(x > value);
+            i = 2 * i + usize::from(x > value);
         }
 
-        // we want ffs(~(i + 1))
-        // since ctz(x) = ffs(x) - 1
-        // we use ctz(~(i + 1)) + 1
-        let j = (i + 1) >> ((!(i + 1)).trailing_zeros() + 1);
-        if j == 0 {
-            None
-        } else {
-            Some(unsafe { self.items.get_unchecked(j - 1) })
-        }
+        // Because the branchless loop navigates the tree until we reach a leaf node regardless of whether
+        // the value is found or not, we now need to decode the found value index, if any.
+        //
+        // To understand how this works, it is useful to think of the index as a binary number.
+        // The index update strategy always multiplies the index by 2 (which can be seen as `i <<= 1`)
+        // and then adds 1 (which can be seen as `i |= 1`) if the value is greater than the current node value.
+        // So, we can interpret the bits in index as a history of turns we made in the tree: a 0 bit means
+        // we went left, a 1 bit means we went right.
+        //
+        // Another important observation is that when we find the target value, we make a left turn
+        // and the corresponding bit in the index will be 0. More importantly, all subsequent bits
+        // will be 1, because after we made a left turn we ended up in a subtree where all values
+        // are less than the target value.
+        //
+        // Therefore, to decode the index we need to:
+        //   1. get rid of all trailing 1 bits (dummy turns we made after we found the target value)
+        //   2. get rid of one more bit to restore the index state before we made a left turn at the target element
+        //   3. check if the resulting index is greater than 0 (0 means the target value is not in the tree)
+        i >>= i.trailing_ones() + 1;
+        (i > 0).then(|| unsafe { self.items.get_unchecked(i).assume_init_ref() })
+    }
+}
+
+impl<T> Drop for OrderedCollection<T> {
+    fn drop(&mut self) {
+        // we need to drop all elements except the 0th one, which is uninitialized
```

Comment:
```suggestion
        // SAFETY: all elements beyond [0] are initialized, so can be dropped (which .truncate(1) will do)
        // at the end of drop(), items will hold a single `MaybeUninit<T>` (`[0]`), which is uninitialized.
        // when the `Vec` is dropped, it will then drop `[0]`, but that's fine since dropping an uninitialized
        // `MaybeUninit<T>` doesn't call `T::drop` and is sound.
```